presto-main/src/main/java/com/facebook/presto/memory/ClusterMemoryLeakDetector.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.memory;
+
+import com.facebook.presto.execution.QueryInfo;
+import com.facebook.presto.execution.QueryState;
+import com.facebook.presto.spi.QueryId;
+import com.google.common.annotations.VisibleForTesting;
+import io.airlift.log.Logger;
+import org.joda.time.DateTime;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.Set;
+import java.util.WeakHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Supplier;
+
+import static com.facebook.presto.execution.QueryState.RUNNING;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static java.util.Collections.newSetFromMap;
+import static java.util.Objects.requireNonNull;
+import static org.joda.time.DateTime.now;
+import static org.joda.time.Seconds.secondsBetween;
+
+@ThreadSafe
+public class ClusterMemoryLeakDetector
+{
+    private static final Logger log = Logger.get(ClusterMemoryLeakDetector.class);
+
+    // It may take some time to remove a query's memory reservations from the worker nodes, that's why
+    // we check to see whether some time has passed after the query finishes to claim that it is leaked.
+    private static final int DEFAULT_LEAK_CLAIM_DELTA_SEC = 60;
```

Comment:
That may cause some queries to be incorrectly marked as leaked. But, in the next run if there are no GCs the state will be built from scratch so those incorrectly marked queries will not be marked. 

I don't think the effect of a full GC is that significant. If we incorrectly mark a query on full GC, the only side effect will be that if the cluster memory manager previously killed that query and is waiting for that query to leave the system, after the full GC it will think that the query is just gone, so it will proceed to kill (if needed).

Do you have a proposal? Maybe we can increase this delta.

Reply:
No proposal, just curious what would happen.  In the future if we want to use this for alerting/monitoring, we may want to make this configurable to reduce false positives.